### ISSUE TITLE
publishDir directive fails on S3 when attempting to copy objects larger than 5 GB

### ISSUE THREAD

**Issue by fstrozzi** [nextflow-io/nextflow#585]
When using the ```publishDir``` directive on a pipeline running on AWS, a COPY operation of an object stored on S3 to the destination specified by the directive fails when objects are larger than 5 GB.

```
WARN  nextflow.processor.PublishDir - Failed to publish file: /nextflow-tmp-fst/8f/1ec21b33cd6e660f74010d18190dcf/table.txt; to: /final/table/table.txt
com.amazonaws.services.s3.model.AmazonS3Exception: The specified copy source is larger than the maximum allowable size for a copy source: 5368709120 (Service: Amazon S3; Status Code: 400; Error Code: InvalidRequest; Request ID: F7CA17FFB876EC35
```

To copy object larger than 5 GB from different S3 locations, the multipart upload API needs to be used:  https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectCOPY.html

**Comment by pditommaso** [comment]
umm, weird multipart upload is supposed to be used.

**Comment by pditommaso** [comment]
Try to enable the logging for the S3 library: 

```
nextflow -trace com.upplication.s3fs run .. etc
```

**Comment by fstrozzi** [comment]
This is the trace corresponding to the error

```
anv.-24 10:54:56.273 [pool-11-thread-1] WARN  nextflow.processor.PublishDir - Failed to publish file: /workdir/44/5e06e6980d39b2b36842eaff52504c/norm_counts.txt; to: /final/tables/test/norm_counts.txt [copy] -- See log file for details
com.amazonaws.services.s3.model.AmazonS3Exception: The specified copy source is larger than the maximum allowable size for a copy source: 5368709120 (Service: Amazon S3; Status Code: 400; Error Code: InvalidRequest; Request ID: 9C71D67F4F2D9329; S3 Extended Request ID: QLYC1r+40qlW54bnXzHfy3oCFnbdTC0SuG2Kfp9aWu1FFnhyfrKgNmqaVRmfKVaOQ213mKK6JUk=)
    at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1587)
    at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1257)
    at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1029)
    at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:741)
    at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:715)
    at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:697)
    at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:665)
    at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:647)
    at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:511)
    at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4227)
    at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4174)
    at com.amazonaws.services.s3.AmazonS3Client.copyObject(AmazonS3Client.java:1854)
    at com.upplication.s3fs.AmazonS3Client.copyObject(AmazonS3Client.java:123)
    at com.upplication.s3fs.S3FileSystemProvider.copy(S3FileSystemProvider.java:539)
    at nextflow.file.FileHelper.copyPath(FileHelper.groovy:828)
    at nextflow.processor.PublishDir.processFileImpl(PublishDir.groovy:306)
    at nextflow.processor.PublishDir.processFile(PublishDir.groovy:281)
    at nextflow.processor.PublishDir.safeProcessFile(PublishDir.groovy:267)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
    at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
    at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1213)
    at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1022)
    at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:925)
    at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:908)
    at org.codehaus.groovy.runtime.InvokerHelper.invokeMethodSafe(InvokerHelper.java:76)
    at nextflow.processor.PublishDir$_apply_closure2.doCall(PublishDir.groovy:240)
    at nextflow.processor.PublishDir$_apply_closure2.call(PublishDir.groovy)
    at groovy.lang.Closure.run(Closure.java:495)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
janv.-24 10:54:58.876 [Task monitor] DEBUG n.processor.TaskPollingMonitor - !! executor awsbatch > tasks to be completed: 1 -- pending tasks are shown belo
```

**Comment by pditommaso** [comment]
Fascinating the publish dir copy, invokes the S3 copy with call directly the s3 [copyObject](https://github.com/nextflow-io/nextflow-s3fs/blob/master/src/main/java/com/upplication/s3fs/S3FileSystemProvider.java#L541-541) which does not implement the multipart upload :/

Need to patch it.

**Comment by fstrozzi** [comment]
ok thanks! :)

**Comment by pditommaso** [comment]
Is this copy from a S3 bucket/path to another S3 bucket/path ?

**Comment by fstrozzi** [comment]
Yes exactly, this is the action triggered by the ```publishDir``` directive in this case

**Comment by pditommaso** [comment]
Requires https://github.com/nextflow-io/nextflow-s3fs/issues/11

**Comment by pditommaso** [comment]
Included in version `0.27.3`
